// reducers/authActions.js
import * as types from '../constants/ActionTypes';
import fetch from 'isomorphic-fetch';
import cookie from 'react-cookie';

export function receiveAuth() {
  const user = cookie.load('username');
  return {
    type: types.AUTH_LOAD_SUCCESS,
    user
  }
}

export function checkAuth() {
  if (cookie.load('username')) {
    return true;
  }
  return false;
}

function requestSignUp() {
  return {
    type: types.AUTH_SIGNUP
  }
}

function receiveUser(username) {
  const newUser = {
    name: username,
    id: Symbol(username)
  }
  return {
    type: types.AUTH_SIGNUP_SUCCESS,
    newUser
  }
}

function requestSignOut() {
  return {
    type: types.AUTH_SIGNOUT
  }
}
function receiveSignOut() {
  return {
    type: types.AUTH_SIGNOUT_SUCCESS
  }
}

export function signOut() {
  return dispatch => {
    dispatch(requestSignOut())
    return fetch('/api/signout')
      .then(response => {
        if(response.ok) {
          cookie.remove('username')
          dispatch(receiveSignOut())
          //browserHistory.push('/')
        }
      })
      .catch(error => {throw error});
  }
}

export function signUp(user) {
  return dispatch => {
    dispatch(requestSignUp())
    return fetch('http://localhost:9010/api/sign_up', {
        method: 'post',
        headers: {},
        body: JSON.stringify(user)
      }).then(response => {
        return response.json();
      }).then(data => {
        console.log("response ", data.token)
        cookie.save('username', user.username)
        cookie.save('token', data.token)
        dispatch(receiveUser(user.username));
      })
      .catch(error => {throw error});
  };
}

function requestSignIn() {
  return {
    type: types.AUTH_SIGNIN
  }
}

function receiveSignIn(token) {
  const user = {
    token: token,
    id: Symbol(token)
  }
  return {
    type: types.AUTH_SIGNIN_SUCCESS,
    user
  }
}

export function signIn(user) {
  console.log("user ", user);
  return dispatch => {
    dispatch(requestSignIn())
     return fetch('http://localhost:9010/api/sign_in', {
      method: 'post',
      headers: { 'Accept': 'application/json', 'Content-Type': 'application/json' },
      body: JSON.stringify(user)
      })
      .then(response => {
        return response.json()
      })
      .then(data => {
        if(data) {
          cookie.save('token', user.token)
          dispatch(receiveSignIn(user.token));
          //browserHistory.push('/chat');
        }
      })
      .catch(error => {throw error});
  };
}

export function receiveSocket(socketID) {
  return {
    type: types.RECEIVE_SOCKET,
    socketID
  }
}
